src/pudl/transform/ferc1.py

```diff
@@ -6745,31 +6752,80 @@ def make_calculation_dimensions_explicit(
     """
     logger.info(f"Adding {dimensions=} into calculation component table.")
     calc_comps_w_dims = calculation_components.copy()
+    on_cols = ["table_name", "xbrl_factoid"]
     # for each dimension, use split/apply/combine. when there are no dims explict in
     # the calc components, merge in all of the dims.
     for dim_col in dimensions:
         # extract the unique observed instances of this one dimension column & add the
         # _calc suffix so we can merge onto the calculation components.
-        observed_dim = (
-            table_dimensions_ferc1[["table_name", "xbrl_factoid", dim_col]]
-            .drop_duplicates()  # bc there are dupes after we removed the other dim cols
-            .rename(
-                columns={
-                    "xbrl_factoid": "xbrl_factoid_calc",
-                    "table_name": "table_name_calc",
-                }
-            )
-        )
+        observed_dim = table_dimensions_ferc1[
+            on_cols + [dim_col]
+        ].drop_duplicates()  # bc there are dupes after we removed the other dim cols
         null_dim_mask = calc_comps_w_dims[dim_col].isnull()
         null_dim = calc_comps_w_dims[null_dim_mask].drop(columns=[dim_col])
         calc_comps_w_implied_dims = pd.merge(
             null_dim,
             observed_dim,
-            on=["table_name_calc", "xbrl_factoid_calc"],
+            on=on_cols,
             how="left",
         )
         calc_comps_w_explicit_dims = calc_comps_w_dims[~null_dim_mask]
         calc_comps_w_dims = pd.concat(
             [calc_comps_w_implied_dims, calc_comps_w_explicit_dims]
         )
     return calc_comps_w_dims
+
+
+def add_parent_dimensions(calc_comps, dimensions):
+    """Add in the dimensions of the parents in the calculation components.
+
+    First, we treat the totals in these dimension columns by assuming that any total we
+    observe is the sum of all non-total dimensions. We implement this by creating new
+    calculation component records which have a parent of dimension total and calculation
+    components of the non-total dimensinos.
+
+    Then, we assume that every parent factoid should have the same dimensions as its
+    calculation component dimensions.
```

Comment:
I think you mean that every parent factoid should have the same dimension **values** as its calculation components? E.g. if a parent has `utility_type=="electric"` then all of its components are also electric? Not that if a parent has a `utility_type` dimension then the child should also have a `utility_type` dimension.

Aren't parent nodes with the value `total` in any dimension an exception to this?

Maybe we mean "With the exception of parents having a value of `total` in some dimension, we assume that all parents should have the same dimension values as the calculation components that they are composed of."

Reply:
Could you add a description of what we expect to be true of the dataframe that's reported at the end of this process?  Or maybe you can explain it to me on the phone and I could write it?

Reply:
i added some context of what these tables looked like coming into this function as well as what actually happens/what is expected on the way out. very happy to chat on the phone about it. i also recommend going to looking at the unit test.